log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -0,0 +1,579 @@
+            }
+            currentOutputFile = null;
+            currentOutputFileName = null;
+            currentOutputStream = null;
+            currentFileSize = 0;
+        }
+
+        // there shouldn't be any queued methods, but be safe
+        queue.clear();
+    }
+
+    private void processQueue()
+    {
+        List<byte[]> batch = new ArrayList<>(MAX_BATCH_COUNT);
+        boolean poisonMessageSeen = false;
+        while (!closed.get() || !poisonMessageSeen) {
+            if (queue.isEmpty()) {
+                try {
+                    batch.add(queue.take());
+                }
+                catch (InterruptedException ignored) {
+                }
+            }
+            else {
+                queue.drainTo(batch, MAX_BATCH_COUNT);
+            }
+
+            int poisonMessageIndex = getPoisonMessageIndex(batch);
+            if (poisonMessageIndex >= 0) {
+                poisonMessageSeen = true;
+                batch = batch.subList(0, poisonMessageIndex);
+            }
+
+            logMessageBatch(batch);
+            batch.clear();
+        }
+    }
+
+    private int getPoisonMessageIndex(List<byte[]> messages)
+    {
+        for (int i = 0; i < messages.size(); i++) {
+            if (messages.get(i) == poisonMessage) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private synchronized void logMessageBatch(List<byte[]> batch)
+    {
+        for (byte[] message : batch) {
+            if (currentFileSize + message.length > maxFileSize) {
+                try {
+                    rollFile();
+                }
+                catch (IOException e) {
+                    // it is possible the roll worked, but there was a problem cleaning up, and it is possible it failed
+                    // either way, roll will be attempted again once file grows by maxFileSize
+                    currentFileSize = 0;
+
+                    reportError("Error rolling log file", e, GENERIC_FAILURE);
+                }
+            }
+
+            logHistoryManager.pruneLogFilesIfNecessary(currentFileSize + message.length);
+
+            currentFileSize += message.length;
+            try {
+                currentOutputStream.write(message);
+            }
+            catch (Exception e) {
+                reportError(null, e, WRITE_FAILURE);
+            }
+        }
+        // always flush at the end of a batch, so logs aren't delayed
+        flush();
+    }
+
+    private synchronized void rollFile()
+            throws IOException
+    {
+        // carefully update the stream, so if there is a problem logging can continue
+
+        LogFileName newFileName = null;
+        Path newFile = null;
+        OutputStream newOutputStream = null;
+        for (int i = 0; i < 100; i++) {
+            try {
+                newFileName = LogFileName.newLogFileName(link, compressionType.getExtension());
```

Comment:
Oh, now I think I see what's going on (I'll leave my comment above for context). I think the problem is that it's not at all clear that `newLogFileName` returns a randomized name. 

Two things contribute to that misunderstanding: 
* the name doesn't suggest that the method does anything other than compose the name and the extension. 
* the method is static, so it's not clear that what state (if any) goes into that randomization.

Reply:
I renamed the method to `generateNextLogFileName` and added the constant `MAX_OPEN_NEW_LOG_ATTEMPTS`